avogadro/io/poscarformat.cpp

```diff
@@ -0,0 +1,337 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+******************************************************************************/
+
+#include "poscarformat.h"
+
+#include <avogadro/core/elements.h> // for atomicNumberFromSymbol()
+#include <avogadro/core/matrix.h> // for matrix3
+#include <avogadro/core/molecule.h>
+#include <avogadro/core/vector.h> // for Vector3
+#include <avogadro/core/unitcell.h>
+#include <avogadro/core/utilities.h> // for split(), trimmed(), lexicalCast()
+
+#include <algorithm> // for std::count()
+#include <iomanip>
+#include <iostream>
+
+namespace Avogadro {
+namespace Io {
+
+using std::getline;
+using std::string;
+using std::vector;
+
+using Core::Array;
+using Core::Atom;
+using Core::Elements;
+using Core::Molecule;
+using Core::UnitCell;
+using Core::lexicalCast;
+using Core::split;
+using Core::trimmed;
+
+PoscarFormat::PoscarFormat()
+{
+}
+
+PoscarFormat::~PoscarFormat()
+{
+}
+
+bool PoscarFormat::read(std::istream &inStream, Core::Molecule &mol)
+{
+  size_t numLines = std::count(std::istreambuf_iterator<char>(inStream),
+                               std::istreambuf_iterator<char>(), '\n');
+
+  // There must be at least 7 "\n"'s to have a minimum crystal (including 1 atom)
+  if (numLines < 7) {
+    appendError("Error: POSCAR file is 7 or fewer lines long");
+    return false;
+  }
+
+  // We have to go back to the beginning if we are going to read again
+  inStream.clear();
+  inStream.seekg(0, std::ios::beg);
+
+  // We'll use these throughout
+  bool ok;
+  string line;
+  vector<string> stringSplit;
+
+  // First line is comment line
+  getline(inStream, line);
+  line = trimmed(line);
+  string title = " ";
+  if (!line.empty()) title = line;
+
+  // Next line is scaling factor
+  getline(inStream, line);
+  double scalingFactor = lexicalCast<double>(line, ok);
+
+  if (!ok) {
+    appendError("Error: Could not convert scaling factor to double in POSCAR");
+    return false;
+  }
+
+  Matrix3 cellMat;
+
+  // Next comes the matrix
+  for (size_t i = 0; i < 3; ++i) {
+    getline(inStream, line);
+    stringSplit = split(line, /*delimiter=*/ ' ');
+    // If this is not three, then there is some kind of error in the line
+    if (stringSplit.size() != 3) {
+      appendError("Error reading lattice vectors in POSCAR");
+      return false;
+    }
+    // UnitCell expects a matrix of this form
+    cellMat(0,i) = lexicalCast<double>(stringSplit.at(0)) * scalingFactor;
+    cellMat(1,i) = lexicalCast<double>(stringSplit.at(1)) * scalingFactor;
+    cellMat(2,i) = lexicalCast<double>(stringSplit.at(2)) * scalingFactor;
+  }
+
+  // Sometimes, atomic symbols go here.
+  getline(inStream, line);
+  stringSplit = split(line, /*delimiter=*/ ' ');
```

Comment:
Most code editors will show you the arguments nowadays, this doesn't hurt but I am not sure it is needed either